File: PrimeNim/solution_1/primes.nim
import times
import math
import strformat
import tables

const DICT = {
    10'u64: 4,
    100'u64: 25,
    1000'u64: 168,
    10000'u64: 1229,
    100000'u64: 9592,
    1000000'u64: 78498,
    10000000'u64: 664579,
    100000000'u64: 5761455,
    1000000000'u64: 50847534,
    10000000000'u64: 455052511,
}.toTable()

type PrimeSieve = ref object of RootObj
    sieveSize: uint64
    bits: seq[bool]

method runSieve(this: PrimeSieve) {.base.} =
    var factor = 3'u64
    let q = uint64(math.sqrt(float64(this.sieveSize)))

    while factor <= q:
        for num in countup(factor, this.sieveSize, 2'u64):
            if not this.bits[num]:
                factor = num
                break

        var num2 = factor * factor
        while num2 < this.sieveSize:
            this.bits[num2] = true
            num2 += (factor * 2'u64)

        factor += 2'u64

method countPrimes(this: PrimeSieve): uint64 {.base.} =
    var count = 1'u64

    for num in countup(3'u64, this.sieveSize, 2'u64):
        if not this.bits[num]:
            count += 1'u64

    return count

method printResults(this: PrimeSieve, showResults: bool, duration: float64, passes: int) {.base.} =
    if showResults:
        stdout.write("2, ")

    var count = 1'u64
    for num in countup(3'u64, this.sieveSize, 2'u64):
        if not this.bits[num]:
            if showResults:
                stdout.write(&"{num}")
            count += 1'u64

    if showResults:
        echo ""

    let
        countPrimes = this.countPrimes()
        isValid = (countPrimes == uint64(DICT[this.sieveSize]))

    stderr.writeLine(&"Passes: {passes}, Time: {duration}, Avg: {(duration / float64(passes))}, Limit: {this.sieveSize}, Count1: {count}, Count2: {countPrimes}, Valid: {isValid}")
    echo &"marghidanu;{passes};{duration};1;algorithm=base,faithful=yes,bits=8"

proc newPrimeSieve(sieveSize: uint64): PrimeSieve =
    PrimeSieve(sieveSize: sieveSize, bits: newSeq[bool](sieveSize))

# --- Main block
var passes = 0
let startTime = times.epochTime()

while true:
    var sieve = newPrimeSieve(1000000'u64)
    sieve.runSieve()

    passes += 1
    var duration = times.epochTime() - startTime
    if duration >= 5:
        sieve.printResults(false, duration, passes)
        break

File: PrimePerl/solution_1/primes.pl
#!/usr/bin/env perl

use strict;
use warnings;

package PrimeSieve {
    my %DICT = (
        10          => 4,
        100         => 25,
        1000        => 168,
        10000       => 1229,
        100000      => 9592,
        1000000     => 78498,
        10000000    => 664579,
        100000000   => 5761455,
        1000000000  => 50847534,
        10000000000 => 455052511,
    );

    sub new {
        my ( $class, $sieve_size ) = @_;

        return bless {
            sieve_size => $sieve_size,
            bits       => [],
        }, $class;
    }

    sub run_sieve {
        my $self = shift;

        my $factor = 3;
        my $q      = sqrt $self->{sieve_size};
        my @bits   = @{$self->{bits}};
        my $ss     = $self->{sieve_size};
        $#bits     = $ss;

        while ( $factor <= $q ) {
            for ( my $num = $factor ; $num < $ss ; $num += 2 ) {
                unless ( $bits[$num] ) {
                    $factor = $num;
                    last;
                }
            }

            my $num2 = $factor ** 2;
            while ( $num2 < $ss ) {
                $bits[$num2] = 1;
                $num2 += $factor * 2;
            }

            $factor += 2;
        }

        $self->{bits} = \@bits;
    }

    sub print_results {
        my ( $self, $show_results, $duration, $passes ) = @_;

        print "2, " if ($show_results);

        my $count = ( $self->{sieve_size} >= 2 );
        for ( my $num = 3 ; $num <= $self->{sieve_size} ; $num += 2 ) {
            unless ( $self->{bits}[$num] ) {
                printf( "%d, ", $num ) if ($show_results);
                $count++;
            }
        }

        print "" if ($show_results);

        printf "marghidanu;%d;%f;%d;algorithm=base,faithful=yes\n", $passes, $duration, 1;
#         printf STDERR
# "Passes: %d, Time: %f, Avg: %f, Limit: %d, Count1: %d, Count2: %d, Valid: %d\n",
#           $passes, $duration, $duration / $passes,
#           $self->{sieve_size}, $count, $self->count_primes(),
#           $self->validate_results();
    }

    sub count_primes {
        my $self = shift;

        my $count = ( $self->{sieve_size} >= 2 );
        for ( my $i = 3 ; $i < $self->{sieve_size} ; $i += 2 ) {
            $count++ unless ( $self->{bits}[$i] );
        }

        return $count;
    }

    sub validate_results {
        my $self = shift;

        return ( $DICT{ $self->{sieve_size} } == $self->count_primes() );
    }
};

package main {
    use Time::HiRes qw(time);

    my $passes     = 0;
    my $start_time = time;

    while (1) {
        my $sieve = PrimeSieve->new(1000000);
        $sieve->run_sieve();
        $passes++;

        my $duration = time - $start_time;
        if ( $duration >= 5 ) {
            $sieve->print_results( 0, $duration, $passes );
            last;
        }
    }
};

__END__
